backend/app/adapter/db/user_changelog.go

```diff
@@ -0,0 +1,98 @@
+package db
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/short-d/short/app/adapter/db/table"
+	"github.com/short-d/short/app/entity"
+	"github.com/short-d/short/app/usecase/repository"
+)
+
+var _ repository.UserChangeLog = (*UserChangeLogSQL)(nil)
+
+// UserChangeLogSQL accesses UserChangeLog information in user_changelog table through SQL.
+type UserChangeLogSQL struct {
+	db *sql.DB
+}
+
+// GetLastViewedAt retrieves LastViewedAt for a given user.
+func (u UserChangeLogSQL) GetLastViewedAt(user entity.User) (time.Time, error) {
+	statement := fmt.Sprintf(`
+SELECT "%s" 
+FROM "%s"
+WHERE "%s"=$1;`,
+		table.UserChangeLog.ColumnLastViewedAt,
+		table.UserChangeLog.TableName,
+		table.UserChangeLog.ColumnEmail,
+	)
+
+	row := u.db.QueryRow(statement, user.Email)
+	lastViewedAt := time.Time{}
+	err := row.Scan(&lastViewedAt)
+	return lastViewedAt, err
+}
+
+// UpdateLastViewedAt updates LastViewedAt time for given user to currentTime.
+func (u UserChangeLogSQL) UpdateLastViewedAt(user entity.User, currentTime time.Time) (time.Time, error) {
+	statement := fmt.Sprintf(`
+UPDATE "%s"
+SET %s=$1
+WHERE %s=$2
+`,
+		table.UserChangeLog.TableName,
+		table.UserChangeLog.ColumnLastViewedAt,
+		table.UserChangeLog.ColumnEmail,
+	)
+
+	result, err := u.db.Exec(
+		statement,
+		currentTime,
+		user.Email,
+	)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	rowsUpdated, err := result.RowsAffected()
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	if rowsUpdated == 0 {
+		return time.Time{}, errors.New("sql: no rows updated")
```

Comment:
I have a proposition:

Let's create declare an error `var ErrRecordNotFound` and return it when there is nothing to update via `UpdateLastViewedAt` and there is nothing to fetch via `GetLastViewedAt`.

We can easily distinguish what kind of error has occurred and to handle it properly. Please look at [errors.Is](https://blog.golang.org/go1.13-errors)

What do you think?

Reply:
We already have a way to deal with this kind of problem by implementing a boolean function to check whether there is an entry in the database (Example is [IsEmailExist](https://github.com/short-d/short/blob/master/backend/app/adapter/db/user.go#L42) in UserSQL), we can implement a similar function `func (u UserChangeLogSQL) UserExists() boolean`  in the UserChangeLog repository. We can call this method in the usecase instead of checking for errors

Reply:
It sounds weird, Let's imagine I got an error by calling function `UpdateLastViewedAt`. 
In order to understand what's happened and if it was `ErrRecordNotFound` error, I should call function `UserExists`. And how should we handle a case when I received an error again on `UserExists` call?

Reply:
@alldroll 
There will be 2 methods in ChangeLog usecase:
1. GetLastViewedAt()
Calls userChangeLogRepo.UserExists() to check if user exists and return nil if user does not exist. Else retrieve last viewed at time using userChangeLogRepo.GetLastViewedAt()

2. UpdateLastViewedAt()
Calls userChangeLogRepo.UserExists() to check if user exists and Create one if not. Else
Update last viewed at time using userChangeLogRepo.UpdateLastViewedAt()

Reply:
LastViewedAt is optional in GraphQL schema: https://github.com/short-d/short/issues/566#issuecomment-602131238

Reply:
Declare error type is Go's way of handling error. Let's do what @alldroll suggested. https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully